doc/developer/arrangements.md

```diff
@@ -0,0 +1,129 @@
+This means that in addition to arranging each of the join inputs, it arranges each of the intermediate results.
+For example, in the figure below, each `A` is an arrangement.
+
+```
+    In2 --> A  In3 --> A  In4 --> A
+             \          \          \
+    In1 --> A-Join --> A-Join --> A-Join --> Out
+```
+
+To join four inputs together, there are six arrangements.
+Some of these input arrangements could be shared, but most likely the intermediate arrangements cannot.
+
+---
+
+Our `Reduce` operator builds a non-trivial dataflow subgraph (which may be subject to change).
+Informally, for aggregation we build its own dataflow.
+That dataflow may begin with a `reduce` operator to implement the `distinct` keyword.
+Then there is either
+    * a single `reduce` operator (for sums, counts, any, all), or
+    * a sequence of 16 `reduce` operators (for min and max), or
+    * a single `reduce` operator (for `jsonbagg` and others).
+Finally, last `reduce` collects each of the aggregates into one record for each key.
+There are special cases when there are few (zero or one) reductions.
+
+---
+
+Our `TopK` operator builds a sequence of 16 `reduce` operators.
+Each `reduce` operator applies the `TopK` logic over progressively coarser groups
+The final group is just by the group key.
+
+---
+
+Our `Threshold` operator creates one `reduce` operator.
+
+---
+
+Our `ArrangeBy` operator creates one arrangement.
+
+---
+
+For each of these relational operators, the associated memory footprint should determined by the sizes of the arranged collections.
+
+## Caveats: Shared Arrangements
+
+Arrangements may be shared between operators in the same dataflow, or across dataflows if an arrangement is published by the dataflow that creates it.
+Arrangements are most commonly published by the creation of indexes, materialized sources, and materialized views.
+Materialize does not currently publish other arrangements.
+
+The `Join` operator requires arrangements each with `key` equal to the columns that will be equated.
+When joining with a collection that has a primary key, we may have that arrangement available.
+In this case, we are able to re-use the existing arrangement, and incur no additional memory use for this arrangement.
+
+The `Reduce` operator concludes with an arrangement of its output, whose `key` is the grouping key.
+It is not uncommon to re-use this arrangement, as we find that such groupings are often followed by joins.
+
+The `mz_arrangement_sharing` logging source reports the number of times each arrangement is shared.
+An arrangement is identified by the worker and operator that created it.
+
+## Caveats: Demand Analysis
+
+When users present queries and views to us, we can determine that some fields are not required.
+We blank out any field that is not required.
+This can reduce the number of distinct `data` in an arrangement, which will reduce the size of the arrangement.
+
+## Caveats: Delta Joins
```

Comment:
Linking to a more comprehensive blog post or reference would be helpful here, in the case that someone wants to learn more.